core/state_transition.libevm_test.go

```diff
@@ -54,3 +60,142 @@ func TestCanExecuteTransaction(t *testing.T) {
+			gasLimit:       1e6,
+			minConsumption: 2e6,
+			wantUsed:       1e6,
+		},
+		{
+			// Although this doesn't test minimum consumption, it demonstrates
+			// the expected outcome for comparison with the next test.
+			name:     "refund_without_min_consumption",
+			gasLimit: 1e6,
+			refund:   1,
+			wantUsed: params.TxGas - 1,
+		},
+		{
+			name:           "refund_with_min_consumption",
+			gasLimit:       1e6,
+			refund:         1,
+			minConsumption: params.TxGas,
+			wantUsed:       params.TxGas,
+		},
+	}
+
+	// Very low gas price so we can calculate the expected balance in a uint64,
+	// but not 1 otherwise tests would pass without multiplying extra
+	// consumption by the price.
+	const gasPrice = 3
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			hooks := &hookstest.Stub{
+				MinimumGasConsumptionFn: func(limit uint64) uint64 {
+					require.Equal(t, limit, tt.gasLimit)
+					return tt.minConsumption
+				},
+			}
+			hooks.Register(t)
+
+			key, err := crypto.GenerateKey()
+			require.NoError(t, err, "libevm/crypto.GenerateKey()")
+
+			stateDB, evm := ethtest.NewZeroEVM(t)
+			signer := types.LatestSigner(evm.ChainConfig())
+			tx := types.MustSignNewTx(
+				key, signer,
+				&types.LegacyTx{
+					GasPrice: big.NewInt(gasPrice),
+					Gas:      tt.gasLimit,
+					To:       &common.Address{},
+					Value:    big.NewInt(0),
+				},
+			)
+
+			const startingBalance = 10 * params.Ether
+			from := crypto.PubkeyToAddress(key.PublicKey)
+			stateDB.SetNonce(from, 0)
+			stateDB.SetBalance(from, uint256.NewInt(startingBalance))
+			stateDB.AddRefund(tt.refund)
+
+			var (
+				// Both variables are passed as pointers to
+				// [core.ApplyTransaction], which will modify them.
+				gotUsed uint64
+				gotPool = core.GasPool(1e9)
+			)
+			wantPool := gotPool - core.GasPool(tt.wantUsed)
+
+			receipt, err := core.ApplyTransaction(
+				evm.ChainConfig(), nil, &common.Address{}, &gotPool, stateDB,
+				&types.Header{
+					BaseFee: big.NewInt(gasPrice),
+					// Required but irrelevant fields
+					Number:     big.NewInt(0),
+					Difficulty: big.NewInt(0),
+				},
+				tx, &gotUsed, vm.Config{},
+			)
+			require.NoError(t, err, "core.ApplyTransaction(...)")
+
+			for desc, got := range map[string]uint64{
+				"receipt.GasUsed":                                  receipt.GasUsed,
+				"receipt.CumulativeGasUsed":                        receipt.CumulativeGasUsed,
+				"core.ApplyTransaction(..., usedGas *uint64, ...)": gotUsed,
+			} {
+				if got != tt.wantUsed {
+					t.Errorf("%s got %d; want %d", desc, got, tt.wantUsed)
+				}
+			}
+			if gotPool != wantPool {
+				t.Errorf("After core.ApplyMessage(..., *%T); got %[1]T = %[1]d; want %d", gotPool, wantPool)
+			}
+
+			wantBalance := startingBalance - tt.wantUsed*gasPrice
+			if got := stateDB.GetBalance(from); !got.IsUint64() || got.Uint64() != wantBalance {
+				t.Errorf("got remaining balance %s; want %d", got.String(), wantBalance)
+			}
```

Comment:
This seems more clear to me:
```suggestion
			wantBalance := uint256.NewInt(startingBalance - tt.wantUsed*gasPrice)
			if got := stateDB.GetBalance(from); !got.Eq(wantBalance) {
				t.Errorf("got remaining balance %d; want %d", got, wantBalance)
			}
```

Reply:
Ahh cool, I didn't realise that `uint256.Int` had a `Format` method so `%d` works! I'm an idiot for not using `Eq()` 🤦